File: 695. Max Area of Island/695. Max Area of Island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        direc=[(0,1),(0,-1),(-1,0),(1,0)]
        m, n, max_area = len(grid), len(grid[0]), 0

        def dfs(r,c):
            grid[r][c]=0
            area=1

            for dr, dc in direc:
                nr=r+dr
                nc=c+dc

                if not (0 <= nr < m and 0 <= nc < n):
                    continue
                
                if grid[nr][nc]==1:
                    area+=dfs(nr,nc)
            return area
        

        for i in range(m):
            for j in range(n):
                if grid[i][j]==1:
                    max_area=max(max_area,dfs(i,j))
        
        return max_area

File: 695. Max Area of Island/695. Max Area of Island.cpp
class Solution {
public:
    vector<pair<int,int>> direc = {
            {0, 1},
            {0, -1},
            {-1, 0},
            {1, 0}
        };
        
    int dfs(vector<vector<int>>& grid, int r, int c){
        int area = 1;
        grid[r][c] = 0;

        for(auto dir : direc){
            int nr = r + dir.first;
            int nc = c + dir.second;

            if (nr < 0 || nr >= grid.size() ||
                nc < 0 || nc >= grid[0].size()) {
                continue;
            }

            if (grid[nr][nc] == 1 ){
                area+=dfs(grid, nr, nc);
            }
        }

        return area;


    }

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int max_area = 0;

        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j< grid[0].size(); j++){
                if (grid[i][j]==1) {
                    max_area=max(max_area,dfs(grid,i,j));
                }
            }
        }

        return max_area;
        
    }
};